keras/layers/core.py

```diff
@@ -589,6 +589,27 @@ def get_config(self):
                 "output_reconstruction": self.output_reconstruction}
 
 
+class Maxout(Layer):
+    '''
+        Max-out layer, nb_feature is the number of pieces in the piecewise linear approx.
+        nb_feature must divide the first dimension of the input.
+    '''
+    def __init__(self, nb_feature=4):
+        super(Maxout, self).__init__()
+        self.nb_feature = int(nb_feature)
+
+    def get_output(self, train=False):
+        X = self.get_input(train)
+        nshape = T.concatenate([(X.shape[0], self.nb_feature, X.shape[1] // self.nb_feature),
+                                X.shape[2:]])
+        output = T.max(T.reshape(X, nshape, ndim=X.ndim+1), axis=1)
```

Comment:
So do you think we should add the dimension on which to perform max as a
parameter?

On Thu, Sep 10, 2015 at 6:44 PM, François Chollet <notifications@github.com>
wrote:

> In keras/layers/core.py
> <https://github.com/fchollet/keras/pull/671#discussion_r39175671>:
>
> > @@ -589,6 +589,27 @@ def get_config(self):
> >                  "output_reconstruction": self.output_reconstruction}
> >
> >
> > +class Maxout(Layer):
> > +    '''
> > +        Max-out layer, nb_feature is the number of pieces in the piecewise linear approx.
> > +        nb_feature must divide the first dimension of the input.
> > +    '''
> > +    def __init__(self, nb_feature=4):
> > +        super(Maxout, self).__init__()
> > +        self.nb_feature = int(nb_feature)
> > +
> > +    def get_output(self, train=False):
> > +        X = self.get_input(train)
> > +        nshape = T.concatenate([(X.shape[0], self.nb_feature, X.shape[1] // self.nb_feature),
> > +                                X.shape[2:]])
> > +        output = T.max(T.reshape(X, nshape, ndim=X.ndim+1), axis=1)
>
> This reshape will fail for some shapes (all inputs X where X.shape[1] is
> odd) because the number of elements in the tensor will not be conserved.
>
> Additionally it appears that the maxout is done only of the 1st tensor
> dimension. This would be time in case of a temporal input, of channels for
> a picture input. I don't thinking that's how it should work in these cases
> (it should be respectively the 2nd and 2nd-3rd dimensions).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/fchollet/keras/pull/671/files#r39175671>.
>